Improve tool validation and metadata generation

mvdbeek opened the pull request:
(Builds on https://github.com/galaxyproject/galaxy/pull/5099)
This introduces a `ValidationConext` object, that should contain just enough information to load tools and data tables and can be used in place of `app`. This avoids manipulation of app-wide settings during tool installations, metadata setting and browsing of toolshed repos.

mvdbeek committed:
Fix copying of loc.sample files

`filename` should be joined with the tool_path, as evidenced by other uses of
copy_sample_file.

mvdbeek committed:
Do not write shed_tool_data_table_config in place

This should prevent errors such as:
```
galaxy.util ERROR 2017-11-17 09:20:12,842 Error parsing file /data/users/mvandenb/gx/config/shed_tool_data_table_conf.xml
Traceback (most recent call last):
  File "lib/galaxy/util/__init__.py", line 217, in parse_xml
    root = tree.parse(fname, parser=ElementTree.XMLParser(target=DoctypeSafeCallbackTarget()))
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 657, in parse
    self._root = parser.close()
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1654, in close
    self._raiseerror(v)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
ParseError: no element found: line 1, column 0
```
reported in https://github.com/galaxyproject/galaxy/issues/5031.

mvdbeek committed:
Fix installation of data manager revision updates

Without this updates of data managers fails with
```
AttributeError: 'NoneType' object has no attribute 'get'
  File "galaxy/web/framework/middleware/sentry.py", line 40, in __call__
    iterable = self.application(environ, start_response)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/paste/recursive.py", line 85, in __call__
    return self.application(environ, start_response)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/paste/httpexceptions.py", line 640, in __call__
    return self.application(environ, start_response)
  File "galaxy/web/framework/base.py", line 136, in __call__
    return self.handle_request(environ, start_response)
  File "galaxy/web/framework/base.py", line 215, in handle_request
    body = method(trans, **kwargs)
  File "galaxy/web/framework/decorators.py", line 98, in decorator
    return func(self, trans, *args, **kwargs)
  File "galaxy/webapps/galaxy/controllers/admin_toolshed.py", line 1849, in update_to_changeset_revision
    persist=True)
  File "tool_shed/galaxy_install/metadata/installed_repository_metadata_manager.py", line 30, in __init__
    metadata_dict=metadata_dict, user=None)
  File "tool_shed/metadata/metadata_generator.py", line 63, in __init__
    self.metadata_dict = {'shed_config_filename': self.shed_config_dict.get('config_filename', None)}
```

mvdbeek committed:
Use a temporary ToolDataTableManager when generating metadata

Otherwise loaded data tables bleed into the app-wide data tables. This can be
seen with lots of data tables pointing to (non-existing, at that point)
temporary files after a metadata reset in the galaxy admin panel,
but it can also cause failure to install tools with entirely new data tables.

mvdbeek committed:
Avoid manipulating global toll data tables

mvdbeek committed:
Avoid global changes to app.config.tool_data_path

mvdbeek committed:
Rename the ToolDataTableManager that deals with repository data to ShedToolDataTableManager to avoid confusion

mvdbeek committed:
Load tools directly from source and avoid loading via toolbox

mvdbeek committed:
Fix missing index file check

mvdbeek committed:
Add miniapp to avoid manipulating global settings

mvdbeek committed:
Always use MiniApp when using ToolValidator

mvdbeek committed:
Only parse citations if app has citations_manager

mvdbeek committed:
Fix linting issues and fix data manager installation

jmchilton commented on lib/galaxy/util/miniapp.py:
```diff
@@ -0,0 +1,70 @@
+from .bunch import Bunch
+
+
+class MiniApp(object):
```
Entirely personal preference - but what about the path and name ``galaxy.tools.repositories.ValidationContext`` instead of ``galaxy.util.miniapp.MiniApp``?

Edit: For more justification - it seems very tool specific - so I don't think it belongs in ``galaxy.util``. I imagine there are other places we will want to decompose ``app`` - indeed there are other places we have already decomposed ``trans`` so we should make it clear what the purpose of a particular decomposition is I guess -for instance with the ``WorkContext`` alternative to ``trans``.

mvdbeek replied:
Sure, I can do that!

mvdbeek committed:
Move and rename MiniApp to ValidationContext

This better describes what it used for and since it is tool specific, fits
better here that in utils. Thanks @jmchilton for the suggestion.

mvdbeek committed:
Move ValidationContext to right path

mvdbeek committed:
Remove more unused variables

mvdbeek committed:
Use ValidationContext data tables instead of separately passing in relevant data tables

mvdbeek committed:
Don't reload tools, new indexes and data tables are recognized without a reload

mvdbeek committed:
Change ValidationContext.build_file_path to be file

Also simplify cleanup code.

mvdbeek committed:
Skip format parsing if datatypes registry contains no datatypes

mvdbeek committed:
Remove unnecessary tool data table resets

mvdbeek committed:
Linting fixes

mvdbeek committed:
Add unit test for ValidationContext and ToolValidator

jmchilton commented:
Amazing, amazing, amazing... amazing! Keep up the amazing work!
